lale/lib/rasl/simple_imputer.py

```diff
@@ -0,0 +1,223 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numbers
+import typing
+
+import numpy as np
+import pandas as pd
+
+import lale.docstrings
+import lale.operators
+from lale.expressions import it, mean, median, mode, replace
+from lale.helpers import _is_df, _is_pandas_df, _is_spark_df
+from lale.lib.sklearn import simple_imputer
+from lale.schemas import Enum
+
+from .aggregate import Aggregate
+from .map import Map
+
+
+class _SimpleImputerImpl:
+    def __init__(
+        self,
+        missing_values=np.nan,
+        strategy="mean",
+        fill_value=None,
+        verbose=0,
+        copy=True,
+        add_indicator=False,
+    ):
+
+        self.missing_values = missing_values
+        self.strategy = strategy
+        self.fill_value = fill_value
+        self.verbose = verbose
+        if not copy:
+            raise ValueError("This implementation only supports `copy=True`.")
+
+        if add_indicator:
+            raise ValueError("This implementation only supports `add_indicator=False`.")
+        # the `indicator_`` property is always None as we do not support `add_indictor=True`
+        self.indicator_ = None
+
+    def fit(self, X, y=None):
+
+        self._validate_input(X)
+
+        # assign appropriate value to fill_value depending on the datatype.
+        # default fill_value is 0 for numerical input and "missing_value"
+        # otherwise
+        if self.fill_value is None:
+            if self.is_numeric_df(X):
+                fill_value = 0
+            else:
+                fill_value = "missing_value"
+        else:
+            fill_value = self.fill_value
+
+        # validate that fill_value is numerical for numerical data
+        if (
+            self.strategy == "constant"
+            and self.is_numeric_df(X)
+            and not isinstance(fill_value, numbers.Real)
+        ):
+            raise ValueError(
+                "'fill_value'={0} is invalid. Expected a "
+                "numerical value when imputing numerical "
+                "data".format(fill_value)
+            )
+
+        # set attribute values
+        self.n_features_in_ = len(X.columns)
+        self.feature_names_in_ = X.columns
+
+        agg_op = None
+        agg_data = None
+        # learn the values to be imputed
+        if self.strategy == "mean":
+            agg_op = Aggregate(
+                columns={c: mean(it[c]) for c in X.columns},
+                exclude_value=self.missing_values,
+            )
+        elif self.strategy == "median":
+            agg_op = Aggregate(
```

Comment:
Do you think we should only support the configurations we can handle with `partial_fit`? I feel we want our operators to be a drop-in replacement of sklearn as much as possible. So even when it is not possible to get the benefits of associativity, I wanted to handle the hyperparameters. And having a spark backend is an additional benefit anyway.

Reply:
Right, I think it's good you implemented all configurations for "fit". For "partial_fit", you can check if the strategy is one of the supported ones, and raise an error otherwise.